### PR TITLE
Search option in watch history and favorite playlist

### LINKED ISSUE #888: Search option in watch history
**Is your feature request related to a usage problem (not a bug)? Please describe.**
Searching for one video in the watch history is quite tedious, especially if it was watched a while back and a lot of other videos filled up the history since then. A search comes in handy when you watched a video that is not easily searchable via normal search. E.g. when you only know the vague theme of the video and the search is filled with a lot of videos about the topic

**Describe the solution you'd like to see implemented**
A search field which allows to search for keywords in titles as well as the channel name of the video. The extend to which data fields can be searched depends on the amount of data that is stored in the history.db file.

### LINKED ISSUE #1403: Search through playlists
**Is your feature request related to a usage problem (not a bug)? Please describe.**
I would like to be able to search through my playlists, as the list can be pretty long.

**Describe the solution you'd like to see implemented**
Maybe a seperate search bar in the playlist window, to seperate it from the main search bar.

**Describe alternatives you've considered**
none

### PR DESCRIPTION
**Pull request by constraintAutomaton**
---
Search option in watch history and favorite playlist
---

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation

**Related issue**
closes #888
closes #1403

**Description**
A search bar is added to the history and playlist pages. The query is searched into the title of the video and the author. Since `nedb` doesn't support text indexing a regex is used to find the videos. Because the documents of the playlists database are the playlist and not the video (which could be refactor for a design where the collections are the playlists and the document are the videos or where the document are the videos with a field that defined the playlists the video is owned) the filtering is made in the front end (we request a whole playlist and filter in the front end the right videos).

**Screenshots**
image

**Testing**
Simply make a query into the search bar and the card should be updated in "realtime"

**Desktop**
 - OS: Ubuntu
- OS Version: 20.04.3 LTS
- FreeTube version: v0.15.1 Beta

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
Just a heads up. Your db/store logic might have to change because I think https://github.com/FreeTubeApp/FreeTube/pull/1833 will be merged before this PR

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> Just a heads up. Your db/store logic might have to change because I think https://github.com/FreeTubeApp/FreeTube/pull/1833 will be merged before this PR

Ready to be reviewed!

**Comment by toby63** [comment]
Very useful PR, thx.

One question, could you maybe also add a search option to the Playlists?

**Comment by constraintAutomaton** [comment]
> Very useful PR, thx.
> 
> One question, could you maybe also add a search option to the Playlists?

I can do it

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Isn't it better to wait for the implementation of playlist to be done?

**Comment by constraintAutomaton** [comment]
> Isn't it better to wait for the implementation of playlist to be done?

 It seems likes it has been started to be implemented in the sense that the database is designed in a way that what is stored is divided in playlist types. But in the front end we only use `Favorites` playlist. My search bar only query into the `Favorites` I'm guessing it wouldn't be difficult to adapt it to others playlists types. (Even though I think it might be better, in my opinion, if the elements of the database were the video instead of the whole playlist .)
 
 But I can delete this feature if you wish

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Ah i see i didnt have the time to look at the WIP implementation of the playlists yet. Thanks for explaining. I think it is ok to keep it in this PR.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Related issue https://github.com/FreeTubeApp/FreeTube/issues/1403

**Comment by toby63** [comment]
Thx for including this for Playlists as well.

I just tried this version and it works very well.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
something goes wrong here

https://user-images.githubusercontent.com/73130443/148604828-ad678aa2-cfff-4388-bd40-502ed91f4317.mp4

**Comment by constraintAutomaton** [comment]
> something goes wrong here
> 
>     1. Cant find the video
> 
>     2. if i clear the query the order of the list has been changed
> 
> 
> VirtualBoxVM_5F91Z3RGA5.mp4

I think this commit fix this issue.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Could i make the implementation of the clear button the same as in this PR https://github.com/FreeTubeApp/FreeTube/pull/1793

This way it will be uniform.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
This happens while using search bar in history section

https://user-images.githubusercontent.com/73130443/148616375-d4c373c4-98c4-47c7-b46f-0ce502de6c65.mp4

https://user-images.githubusercontent.com/73130443/148616378-93686e0d-053f-4999-a2d9-54f6f3dd2113.mp4

**Comment by constraintAutomaton** [comment]
> Could i make the implementation of the clear button the same as in this PR #1793
> 
> This way it will be uniform.

since I'm using the same component, I think when it is merged the style will be unified

**Comment by constraintAutomaton** [comment]
I think I've fixed those issues.

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM!

**Comment by constraintAutomaton** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc I wonder if you can request some reviews for this pr?

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Sure! :)

**Review by PikachuEXE — COMMENTED** [review]
Reviewed the code
Will test the feature later

**Review by PikachuEXE — COMMENTED** [review]
I have tested both watch history and favorite playlist
I only have 1 question before approving:
How would someone know/want to search a video via videoId instead of URL???

**Comment by constraintAutomaton** [comment]
> I have tested both watch history and favorite playlist I only have 1 question before approving: How would someone know/want to search a video via videoId instead of URL???

good point I'm gonna support the url (by cutting it to get the id)

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
im with @constraintAutomaton on this one it doesnt really make sense to add the url because if u have the url that means u have the title of the video so why would u want both?

**Review by PikachuEXE — APPROVED** [review]
I am fine with searching with video ID (no negative effect I think)
Just not sure who would obtain a video ID just to search video in FT normally...

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

File: src/datastores/index.js [hunk 1]
```diff
@@ -18,4 +18,7 @@ db.profiles = Datastore.create({ filename: dbPath('profiles'), autoload: true })
 db.playlists = Datastore.create({ filename: dbPath('playlists'), autoload: true })
 db.history = Datastore.create({ filename: dbPath('history'), autoload: true })
 
+db.history.ensureIndex({ fieldName: 'author' })
+db.history.ensureIndex({ fieldName: 'title' })
+db.history.ensureIndex({ fieldName: 'videoId' })
```
Comment (ChunkyProgrammer) [inline src/datastores/index.js:23]: Don't think this is needed anymore

**Review by ChunkyProgrammer — APPROVED** [review]

**Comment by Svallinn** [comment]
Did nobody notice the obviously unnecessary duplicated datastores file (`src/renderer/store/datastores.js`) that this PR added???

**Comment by PikachuEXE** [comment]
@Svallinn 
I only notice it after your comment
It should be merged with `src/datastores/index.js`?

**Comment by Svallinn** [comment]
No, the code comes from an old file that used to exist, but it's not necessary anymore. Just delete it.

**Comment by PikachuEXE** [comment]
I think this PR relying on code like
```js
historyDb.ensureIndex({ fieldName: 'author' })
historyDb.ensureIndex({ fieldName: 'title' })
historyDb.ensureIndex({ fieldName: 'videoId' })
```

**Comment by PikachuEXE** [comment]
@Svallinn #2023 opened

**Pull request merged**